base_layer/core/src/proof_of_work/diff_adj_manager/diff_adj_storage.rs

```diff
@@ -0,0 +1,208 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    blocks::blockheader::BlockHash,
+    chain_storage::{BlockchainBackend, BlockchainDatabase},
+    proof_of_work::{
+        diff_adj_manager::error::DiffAdjManagerError,
+        difficulty::DifficultyAdjustment,
+        lwma_diff::LinearWeightedMovingAverage,
+        Difficulty,
+        PowAlgorithm,
+        ProofOfWork,
+    },
+};
+use tari_transactions::consensus::DIFFICULTY_BLOCK_WINDOW;
+use tari_utilities::{epoch_time::EpochTime, hash::Hashable};
+
+/// The UpdateState enum is used to specify what update operation should be performed to keep the difficulty adjustment
+/// system upto date with the blockchain db.
+enum UpdateState {
+    FullSync,
+    SyncToTip,
+    Synced,
+}
+
+/// DiffAdjManager makes use of DiffAdjStorage to provide thread save access to its LinearWeightedMovingAverages for
+/// each PoW algorithm.
+pub struct DiffAdjStorage<T>
```

Comment:
This is a lot of code for a "keep a running list of the last N block data"-- it works, so we'll leave it as is, but did you consider any other strategies?:

* Use VecDequeues and just update from the DB every time? i.e. assume we're always in `SyncToTip` but you pop off the front and push to the back until you have enough data
* Or make it someone else's job to keep the Queues up to date (like a subscriber to a newBlock message)